### PR TITLE
fix(core-amqp): leave an object closed when its close fails

### LINKED ISSUE #7323: [azure-core-amqp] A close that fails stops the process, because the object stays open
## Summary

A close that fails stops the process. `ManagementClientImpl::Close` closes its message sender and its message receiver before it clears its own open flag (`sdk/core/azure-core-amqp/src/impl/uamqp/amqp/management.cpp:290-319`). The sender close throws when the context of the caller is cancelled or when its deadline has passed. The flag then stays set, and `ManagementClientImpl::~ManagementClientImpl` calls `AzureNoReturnPath` (`management.cpp:32-40`). `MessageSenderImpl` and `MessageReceiverImpl` hold the same abort in their own destructors (`message_sender.cpp:124-125`, `message_receiver.cpp:256-257`). A user cannot catch any of these, because the process stops.

## Motivation

`MessageSenderImpl::Close` waits on its close queue with the context of the caller. `WaitForResult` gives no result when that context is cancelled, and the function throws `OperationCancelledException` (`sdk/core/azure-core-amqp/src/impl/uamqp/amqp/message_sender.cpp:415-419`). `ManagementClientImpl::Close` calls that close at `management.cpp:305`, and it sets `m_isOpen = false` only at `management.cpp:318`. So the throw skips the assignment and leaves the client open.

A caller of the Event Hubs library reaches this code with its own context. `ProducerClient::GetEventHubProperties` takes an `Azure::Core::Context` (`sdk/eventhubs/azure-messaging-eventhubs/inc/azure/messaging/eventhubs/producer_client.hpp:170`). That context goes to `ManagementClientImpl::ExecuteOperation`, then to `ConnectionImpl::AuthenticateAudience`, then to `PutTokenForAudience` in `sdk/core/azure-core-amqp/src/amqp/connection.cpp`, which closes a claims based security object with the same context. So a caller that puts a deadline on a properties call can stop the process.

The close queue completes in two places only: the change of state from closing to idle (`message_sender.cpp:283-286`) and the detach handler (`message_sender.cpp:476`). A sender that goes to the error state completes the open queue or the send queue instead. So a peer that sends no detach also makes the close throw, and that gives the same abort.

A deadline on the close is not a correction. A test gave the close its own context with a deadline of ten seconds. The Event Hubs concurrency test then stopped the process on each run, at the same assert, because a close under load needs more than ten seconds. The same test passes in 63 seconds when the close keeps the context of the caller. A deadline turns a close that is slow into a process that stops.

Issue #7254 and pull request #7308 both touch this code. Pull request #7308 records this defect in its follow up section, and it does not correct it.

## Proposal

Make a close that fails leave the object closed, because an object that failed to detach is not usable again.

1. Clear `m_isOpen` in `ManagementClientImpl::Close` on every path, including the paths that throw. A scope guard does this, and it keeps the current order of the calls.
2. Clear `m_senderOpen` in `MessageSenderImpl::Close` and `m_receiverOpen` in `MessageReceiverImpl::Close` in the same way. All three flags need this, because each class stops the process on its own.
3. Let the exception go to the caller after the flags are clear. A caller can then see that the close failed, and the destructor no longer stops the process.
4. Add tests that close an object with a context that is already cancelled, and then destroy that object. The test must complete. Today it stops the test process.

Do not add a deadline to the close. The paragraph above gives the measurement that rejects that approach.

### PR DESCRIPTION
**Pull request by j7nw4r**
## Summary

`ManagementClientImpl::Close`, `MessageSenderImpl::Close`, and `MessageReceiverImpl::Close` clear the open flag only on the last line. A close that throws skips that line and leaves the object open. The destructor then calls `AzureNoReturnPath`, which stops the process in a Debug build and in a Release build. The caller cannot catch this.

Fixes #7323.

## Motivation

`MessageSenderImpl::Close` waits on its close queue with the context of the caller. A cancelled context, or a peer that sends no detach, makes `WaitForResult` return nothing. The close then throws `OperationCancelledException`. An Event Hubs caller that puts a deadline on a properties call therefore stops the process.

The uAMQP sender and receiver also run their teardown after that wait, so a throw skips it. The teardown calls `EnableAsyncOperation(false)`. Without that call the poll count stays raised, and `~ConnectionImpl` aborts with "Connection is being destroyed while polling". A fix that only clears the flag moves the abort. `ManagementClientImpl::Close` skips its receiver close when the sender close throws, and the open receiver aborts in its destructor.

Pull request #7308 rebases on this change. Issue #7327 records the one open item in this area.

## Changes

- `MessageSenderImpl` and `MessageReceiverImpl` (uAMQP) move the teardown into a private `CompleteClose` method. `Close` calls it on the normal path and in a catch block that rethrows. `CompleteClose` releases the link, calls `EnableAsyncOperation(false)`, and clears the open flag.
- `ManagementClientImpl::Close` (uAMQP) runs both closes, keeps the first exception, clears its own flag, and rethrows.
- `ManagementClientImpl::Close` and `MessageSenderImpl::Close` (Rust) clear the flag before the detach call. This matches the Rust `MessageReceiverImpl::Close`, which already did this.
- No close gets its own deadline. A ten second deadline stopped the Event Hubs concurrency test on each run. That test passes in 63 seconds when the close keeps the context of the caller.
- The changelog gets an entry under `### Bugs Fixed` for 1.0.0-beta.13.

## Test plan

Three tests open an object, close it with a context that is already cancelled, and then destroy it.

- `TestMessageSendReceive.SenderCloseWithCancelledContext` and `TestMessageSendReceive.ReceiverCloseWithCancelledContext` in `sdk/core/azure-core-amqp/test/ut/message_sender_receiver.cpp`. Both register a mock service endpoint. Both assert `EXPECT_THROW(..., Azure::Core::OperationCancelledException)` on uAMQP and `EXPECT_NO_THROW` on Rust.
- The receiver test also waits for the Open state through a `MessageReceiverEvents` handler. `MessageReceiverImpl::Open` returns before the link attaches, so an immediate close leaves `shouldWaitForClose` false and never reads the context.
- `TestManagement.ManagementCloseWithCancelledContext` in `sdk/core/azure-core-amqp/test/ut/management_tests.cpp` asserts the same exception. Its body sits behind `#if !defined(USE_NATIVE_BROKER)`, so it is empty on a Rust build.

All three sit inside the `#if !defined(AZ_PLATFORM_MAC)` block that holds the tests around them.

The pipeline builds the Rust stack alone, because `sdk/core/azure-core-amqp/CMakeLists.txt` forces `USE_RUST_AMQP` on unless `DISABLE_RUST_IN_BUILD` is set. A cancelled context cannot make a Rust detach fail. `amqpmessagesender_detach_and_release` in `rust_wrapper/src/amqp/message_sender.rs` uses the call context only for the runtime handle and the error slot, then calls `block_on(sender.inner.detach())`.

## Validation

A Linux container ran the tests on the branch, then ran them again with the production code of `origin/main` restored.

| Test | With the change | With the production code of `main` |
| --- | --- | --- |
| `TestMessageSendReceive.SenderCloseWithCancelledContext` | exit 0, `[ OK ] ... (510 ms)` | exit 134, `AzureNoReturnPath (msg="MessageSenderImpl is being destroyed while open.")` at `message_sender.cpp:125` |
| `TestMessageSendReceive.ReceiverCloseWithCancelledContext` | exit 0, `[ OK ] ... (516 ms)` | exit 134, `AzureNoReturnPath (msg="MessageReceiverImpl is being destroyed while open.")` at `message_receiver.cpp:257` |
| `TestManagement.ManagementCloseWithCancelledContext` | exit 0, `[ OK ] ... (720 ms)` | exit 134, `management.cpp:37: ~ManagementClientImpl(): Assertion ((void)("Management being destroyed while open."), (!m_isOpen)) failed` |

- Each abort site above comes from a `gdb` backtrace of the aborting run, so the frame and the message text are exact.
- The full uAMQP suite passes with the change: `[ PASSED ] 167 tests`, exit 0. With the production code of `main` it reaches 120 tests, then aborts at `TestManagement.ManagementCloseWithCancelledContext` with exit 134.
- The build: `ubuntu:22.04`, GCC 11.4.0, CMake 4.3.4 from the Kitware release (vcpkg needs `string(JSON ... STRING_ENCODE)` and Ubuntu 22.04 ships CMake 3.22), and vcpkg at the `builtin-baseline` of `vcpkg.json`. Then `cmake -S . -B build-uamqp -G Ninja -DCMAKE_TOOLCHAIN_FILE=<vcpkg>/scripts/buildsystems/vcpkg.cmake -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTING=ON -DDISABLE_RUST_IN_BUILD=ON` and `cmake --build build-uamqp --target azure-core-amqp-tests`. The build needs no `WARNINGS_AS_ERRORS=OFF`.
- The container is `linux/arm64` and the pipeline uses `x86_64`. The defect sits in C++ control flow and not in code generation, but this measurement does not cover `x86_64`.
- The pipeline covers the Rust stack, which this measurement did not build.

### CODE REVIEW

File: sdk/core/azure-core-amqp/test/ut/message_sender_receiver.cpp [hunk 1]
```diff
@@ -354,6 +354,200 @@ namespace Azure { namespace Core { namespace Amqp { namespace Tests {
...
+      // The close can fail, because the context is cancelled. The caller gets the exception, and
+      // the receiver goes to the closed state.
+      try
+      {
+        receiver.Close(cancelledContext);
+      }
+      catch (std::exception const& e)
+      {
+        GTEST_LOG_(INFO) << "Close threw, which the cancelled context allows: " << e.what();
+      }
```
[the remark this answers is not in the document]
Comment (j7nw4r) [reply sdk/core/azure-core-amqp/test/ut/message_sender_receiver.cpp]: Good catch. Both tests caught every exception, so they passed with or without the rethrow.

They now assert `OperationCancelledException` on uAMQP, where the close waits for the detach on the caller's context. The Rust transport keeps the no-throw expectation, because `CallContext` does not pass the cancellation to `amqpmessagesender_detach_and_release` or `amqpmessagereceiver_detach_and_release`.

Fixed in e4ad14e.

File: sdk/core/azure-core-amqp/test/ut/management_tests.cpp [hunk 2]
```diff
@@ -298,6 +298,54 @@ namespace Azure { namespace Core { namespace Amqp { namespace Tests {
...
+      // The close can fail, because the context is cancelled. The caller gets the exception, and
+      // the client goes to the closed state.
+      try
+      {
+        management.Close(cancelledContext);
+      }
+      catch (std::exception const& e)
+      {
+        GTEST_LOG_(INFO) << "Close threw, which the cancelled context allows: " << e.what();
+      }
```
[the remark this answers is not in the document]
Comment (j7nw4r) [reply sdk/core/azure-core-amqp/test/ut/management_tests.cpp]: Same problem as the receiver and sender tests, and the same fix.

The test now asserts `OperationCancelledException`. `ManagementClientImpl::Close` closes the message sender first, and that close waits for the detach on the caller's context, so the sender's exception becomes `firstException` and reaches the caller. A close that swallows the failure, or that replaces it with the receiver's exception, now fails the test.

The test body needs no transport guard, because `USE_NATIVE_BROKER` removes it on the Rust transport.

Fixed in bf39a3d.

**Comment by j7nw4r** [comment]
Closing this one. The change already reached `main` through #7308, which merged as c6c3f94c9. That branch sat on top of this commit, so the merge carried this fix in as its first commit.

Everything here is in `main` today: `CompleteClose` in the uAMQP sender and receiver, the first-exception handling in the uAMQP management client, the flag-before-detach on the Rust side, all three `EXPECT_THROW(..., OperationCancelledException)` tests, and the changelog bullet. #7323 is closed. The only diff left against `main` is a duplicate of that same changelog bullet, which is what marks this pull request as conflicting.
